### ISSUE TITLE
Search results can include nodes that aren't in the OZ tree

### ISSUE THREAD

**Issue by jaredkhan** [OneZoom/OZtree#879]
## Expected
Search results on the explorer on onezoom.org should all be clickable and take you to a place on the tree

## Actual
Some results don't have a place on the tree and show an error when clicked.
Particular examples are when searching for 'Foxes' or 'Flying foxes'

https://github.com/user-attachments/assets/3b2b610b-3867-4fb9-82ea-40b41b8cbc6e

**Comment by hyanwong** [comment]
Hmm, this is very well observed with "foxes". We currently have 3 OTTs with the exact vernacular "foxes": 770319, 821970, and 372706. These all exist in ordered_nodes.

The problem seems to be that the first hit in the search term is `Clade1634_`, and thereafter, all the links (even the ones labelled "Vulpes", are to `https://www.onezoom.org/life/@Clade1634`, e.g. `<a href="https://www.onezoom.org/life/@Clade1634">Vulpes</a>`. I wonder if we are doing something weird with the trailing underscore somehow?

**Comment by hyanwong** [comment]
Hmm, and now I can't seem to get this to replicate. It looks like once the names are cached, it all works. This is very strange.

**Comment by hyanwong** [comment]
Ah, it looks like when you first go to a search hit, the resulting URL is then populated to all the items in the search. Here's some source code that was served up to me (note all the links are going to the wrong place). Do you have any idea what's going on here @lentinj ?
<img width="514" alt="Screenshot 2024-09-12 at 19 02 49" src="https://github.com/user-attachments/assets/df84904f-7f0e-45a9-a05f-ab38f02b7955">

**Comment by lentinj** [comment]
Found it! It's not a caching issue, it's the fact that the current page URL is ``https://www.onezoom.org/life/@Clade3272``. Navigate to that, you get an error, then start searching.

When populating the search results, there's an attempt to mangle the URL to make sure we have the right ``life``:

https://github.com/OneZoom/OZtree/blob/118bfba9b122d812386e438fb6d6b29cb51aed7e/OZprivate/rawJS/OZTreeModule/src/ui/search.js#L12-L18

Have a URL without a ``=`` in, and that regex fails, resulting in all the search links going to the current URL.

However this mangling, as well as being broken, is probably pointless. As discussed in #870 recently, ``closeAndLeap()`` strips any ``/life/`` before navigating:

https://github.com/OneZoom/OZtree/blob/889026e68377ae306e8f3ffc60e16223803420cc/views/treeviewer/layout.html#L226-L227

So ditching lines 12--15 seems reasonable. Maybe we should try harder to generate the right links so you can open search results in a new tab, but this isn't the way to go about it anyway.

**Comment by hyanwong** [comment]
Well found @lentinj. I'm happy to go with whatever you suggest. I wonder if there's any way to set up tests for this sort of thing, though?

**Comment by lentinj** [comment]
You could use JSDOM to wrap enough of a dummy document to give ``searchPopulate`` a searchbox, then compare the ``searchbox.innerHTML`` once it's done. There's similar setup for the tour tests:

https://github.com/OneZoom/OZtree/blob/118bfba9b122d812386e438fb6d6b29cb51aed7e/OZprivate/rawJS/OZTreeModule/tests/util_tourwrapper.js#L28-L44

Longer-term I'd like to wrap up the search box as a web component, creating more of a defined API for adding them to the page (and to test), but we're not there yet.

**Comment by jaredkhan** [comment]
Just jotting down another case of this: the top result for 'Tortoises'

**Comment by davidebbo** [comment]
> Just jotting down another case of this: the top result for 'Tortoises'

Interesting related notes:
- the display string when searching is "Tortoises (Clade591_)"
- manually navigating to the Tortoise node doesn't find an OTT for it
- Searching for Testudinidae doesn't find anything

Would need to dig into the tree to see what's happening.

**Comment by jaredkhan** [comment]
Another case of this: the top result for 'Big Cats'

Text in the search dropdown is 'Big Cats (Clade7728_)'
Link on the tag is https://www.onezoom.org/life/@Clade7728?otthome=%40_ozid%3D1

**Comment by davidebbo** [comment]
> Another case of this: the top result for 'Big Cats'
> 
> Text in the search dropdown is 'Big Cats (Clade7728_)' Link on the tag is https://www.onezoom.org/life/@Clade7728?otthome=%40_ozid%3D1

I think this happens when to (pseudo) taxon name has an underscore. In this case, the name is `Clade7728_`, and that throws something off. Probably some kind of escape issue, though I have not digged into it.

**Comment by lentinj** [comment]
It's the lack of equals. See https://github.com/OneZoom/OZtree/issues/879#issuecomment-2348465417

**Comment by davidebbo** [comment]
Though https://www.onezoom.org/life/@Panthera works with no equal

**Comment by lentinj** [comment]
By the time you're displaying search results it's become https://www.onezoom.org/life/@Panthera=563154, whereas https://www.onezoom.org/life/@Clade7728 is still https://www.onezoom.org/life/@Clade7728

**Comment by davidebbo** [comment]
Ok, but then let's look at another one that has a name but no OTT: https://www.onezoom.org/life/@Podiata. This one doesn't fail. In the end, it does end up on a different node: https://www.onezoom.org/life/@Eukaryota=304358.

What explains the different behavior between `Podiata` and `Clade7728_`? I think it is the underscore. In fact, if I rename `Podiata` to `Podiata_` in my DB, it becomes broken like `Clade7728_`.

**Comment by hyanwong** [comment]
I use trailing underscores to indicate that the name is not a valid scientific name, but simply a OneZoom convenience. As such, names with a trailing underscore should not appear in searches.

**Comment by hyanwong** [comment]
Ah, sorry, I mean the name itself shouldn't be displayed, but any vernaculars could be.

**Comment by jaredkhan** [comment]
The URL re-writing problem is maybe worth fixing but is a separate problem imo, unless I'm misunderstanding. That was only going on because we got to an invalid pinpoint in the first place. 

I've opened a little PR to avoid making bad pinpoints (e.g. @"Clade7728") for these fake-named nodes (e.g. "Clade7728_") which fixes the cases mentioned here like 'Foxes', 'Big Cats', and 'Tortoises'. See what you think:
https://github.com/OneZoom/OZtree/pull/927
